[SAMVEG] Custom case upload

mkangia opened the pull request:
:fish: 
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
First set of changes to set up custom case upload for SAMVEG through extensions.
The changes are only applicable to `SAMVEG_DOMAINS` which is a collection of domains saved in the code itself.

![Screenshot from 2022-01-12 18-29-54](https://user-images.githubusercontent.com/3864163/149145150-ec8e9e23-276e-40bc-80cf-e0bd98bb8723.png)

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Tech Spec](https://docs.google.com/document/d/1OYrKoJbuEdUrt4M46k3LMiUz6rjnevLx6BsLJ3eCmzs/edit#heading=h.j30o2593mn66)

Using extensions to put in custom operations to perform
1. Custom validations on the file
2. Custom validations on each case upload row
3. Custom operations on each case upload row

Used extensions instead of feature flag for cleaner code interface.

This is the parent PR. Takes care of
1. [Mandatory fields](https://docs.google.com/document/d/1OYrKoJbuEdUrt4M46k3LMiUz6rjnevLx6BsLJ3eCmzs/edit#heading=h.afqz8s4au33s)
2. [Call Checks](https://docs.google.com/document/d/1OYrKoJbuEdUrt4M46k3LMiUz6rjnevLx6BsLJ3eCmzs/edit#heading=h.xt3and1gtjfb)

Rest in follow up PRs.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Tested locally that things are working fine for usual domains. This is also undergoing QA.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Adding tests for new set of changes.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3802


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
hooks custom case upload operations via extensions

mkangia committed:
validates mandatory columns

mkangia committed:
validates mandatory values per row

mkangia committed:
adds missing init file

mkangia committed:
updates to better name

mkangia committed:
removes redundant message

mkangia committed:
adds default return

mkangia committed:
fixes method by always returning a list

mkangia committed:
have the same message for better UI

mkangia committed:
validates call columns

- at least one present
- latest one value in last month

mkangia committed:
limits upload per owner per call

mkangia committed:
adds new argument

mkangia commented:
Hey @snopoke @orangejenny @millerdev 
the timeline for this is a bit tight so if you could take a first pass at this and the follow up PR today, that would help keep up with timeline. Hopefully this should not take long.

millerdev commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -118,6 +119,10 @@ def do_import(self, spreadsheet):
+        fields_to_update, error_messages = custom_case_import_operations(self.domain, row_num, raw_row,
+                                                                         fields_to_update)
+        if error_messages:
+            raise exceptions.CustomCaseRowError(message=", ".join(error_messages))
```
Some error messages contain commas, if I am not mistaken. Would it make sense to use a semicolon as a delimiter here?
```suggestion
            raise exceptions.CustomCaseRowError(message="; ".join(error_messages))
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/30921/commits/e86cf2c56b1c67d2ab86f04580cca5ddfc7a1d76

orangejenny reviewed (commented):
My biggest feedback is that it'd be better to store structured row-level errors. I don't think the current code is going to work with the recent improvements to case importer error messaging:

![Screen Shot 2022-01-11 at 1 58 11 PM](https://user-images.githubusercontent.com/1486591/149004545-04f36cbd-d5fb-4de5-9692-0c8cc3de19c3.png)

millerdev commented on corehq/apps/case_importer/extension_points.py:
```diff
@@ -0,0 +1,28 @@
+    Parameters:
+        Domain name, row number, raw excel row, fields_to_updated
+    Returns:
+        Final fields to update, error messages
```
```suggestion
        Tuple: dict of fields to update, list of error messages
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/30921/commits/b636b3d57fc56281a383259b9996521dc229e0e5

mkangia committed:
fetches directly from list

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

mkangia committed:
makes extra arg optional

mkangia committed:
updates key for better readability

mkangia committed:
remove redundant var

mkangia commented on corehq/apps/case_importer/extension_points.py:
```diff
@@ -0,0 +1,14 @@
+
+
+@extension_point(result_format=ResultFormat.FLATTEN)
+def custom_case_upload_file_operations(domain, case_upload) -> List[str]:
```
> Doesn't case importer report errors per row, and store those errors in a structured format?

It combines errors that have the same message and show "Row(s)" under the message then.

> but doesn't this function need a body?

The function should have a body if there is a default implementation needed or if you see the next method, it returns the fields_to_update and an empty list as its default implementation. If there is no definition, `None` is returned.

orangejenny replied:
Interesting, I didn't know you could leave off the body.

mkangia replied:
> It combines errors that have the same message

Correction that HQ uses error titles to combine errors and not messages. Title is set on the class, so its effectively grouping by classes.
https://github.com/dimagi/commcare-hq/blob/44f54201134f654a46ce05c6574b654600aad87e/corehq/apps/case_importer/do_import.py#L547

for the message, it just uses the last message passed for the kind of error, which the user sees
https://github.com/dimagi/commcare-hq/blob/44f54201134f654a46ce05c6574b654600aad87e/corehq/apps/case_importer/do_import.py#L553

mkangia commented on corehq/apps/case_importer/views.py:
```diff
@@ -152,6 +159,10 @@ def _process_file_and_get_upload(uploaded_file_handle, request, domain, max_colu
 
+    error_messages = custom_case_upload_file_operations(domain=domain, case_upload=case_upload)
+    if error_messages:
+        raise CustomImporterError(", ".join(error_messages))
```
These would be the errors for the file. For now, this would be a single error only i.e the list would have only a single error but has been kept a list to add more errors/checks in future.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -0,0 +1,12 @@
+        raise NotImplementedError
+
+
+class MandatoryColumnsValidator(BaseValidator):
```
Ya. Those PRs are still up but the project needs this done this week itself.

On that, next step is for me to right specs and then share with USH and then it would be considered further. I am guessing once this is introduced in DD, we can remove code from here and use DD instead or may be just keep these. Will depend on how things are once we get to that point.

mkangia commented on custom/samveg/const.py:
```diff
@@ -1 +1,24 @@
+NEWBORN_WEIGHT_COLUMN = 'newborn_weight'
+
+MANDATORY_COLUMNS = [
+    'name',
```
Yup, these are from an external system.

mkangia commented on custom/samveg/tests/case_importer/validators.py:
```diff
@@ -0,0 +1,43 @@
+    root = os.path.dirname(__file__)
+
+    def test_unexpected_sheet(self):
+        with get_spreadsheet(self.get_path('unexpected_sheet_case_upload', 'xlsx')) as spreadsheet:
```
Ya, in app translations that definitely made it easier to see tests.
The code we are testing expects a spreadsheet to be passed, so I would need to create a workbook in memory and then pass it to `get_spreadsheet` which replicates how case upload reads the files? I will try that, may be take that as a follow up.

orangejenny replied:
It'd be a useful test utility to transform tuples into an xls (if we don't already have one). But yeah, that seems fine as a followup.

mkangia commented on corehq/apps/case_importer/extension_points.py:
```diff
@@ -0,0 +1,28 @@
+    Parameters:
+        Domain name, row number, raw excel row, fields_to_updated
+    Returns:
+        Final fields to update, error messages
```
I already added the skip part in the comment above, this was just meant to show what is returned from this function.
I just took this from the convention from other extension methods.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -38,6 +38,32 @@ def _validate_missing_columns(cls, spreadsheet):
+    def run(cls, row_num, raw_row, fields_to_update):
+        error_messages = []
+        error_messages.extend(cls._validate_mandatory_columns(row_num, raw_row))
+        return fields_to_update, error_messages
```
not sure what you mean here @orangejenny. The row would be skipped eventually in core HQ, if there are error messages returned from extension.

orangejenny replied:
I think I misunderstood the expected return value of this function.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -38,6 +38,32 @@ def _validate_missing_columns(cls, spreadsheet):
+        columns = set(raw_row.keys())
+        mandatory_columns = get_mandatory_columns(columns)
+
+        for mandatory_column in mandatory_columns:
```
Ya, I did use set earlier but since the order in set could change, the UI would have 2 different error messages for the same set of missing fields. Like if 2 rows have missing A and B column, they might end up with messages as "Missing values for A, B" and "Missing values for B, A", leading to two separate entries in UI.
So to have predictable order I had to use lists.

mkangia commented on custom/samveg/commcare_extensions.py:
```diff
@@ -1,18 +1,37 @@
 ]
 
 
+row_level_operations = [
```
I will be happy to update this to uppercase. I was thinking of this more like a module-level var rather than module-level constant.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -0,0 +1,12 @@
+class BaseValidator:
+    @classmethod
+    def run(cls, *args, **kwargs):
```
Yes, the signature can be different for different validators.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -0,0 +1,117 @@
+        raise NotImplementedError
+
+    @classmethod
+    def _get_latest_call_value_and_number(cls, raw_row):
```
Sorry yes, I already knew it would be called later in https://github.com/dimagi/commcare-hq/pull/30928/files#diff-e2ad39182718855af1cb8edaa1ad795f29b9c18d7acca6e5db02fb8791eb57b8R118, so I had it in the base class.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -0,0 +1,117 @@
+        latest_call_number = None
+        for i in range(1, 7):
+            if raw_row.get(f"Call{i}"):
+                latest_call_value = raw_row.get(f"Call{i}")
```
Seems like one commit is present in this PR and also in the child PR.
This was addressed in https://github.com/dimagi/commcare-hq/pull/30928#discussion_r782410353

mkangia commented:
> I don't think the current code is going to work with the recent improvements to case importer error messaging:

Alright. I will take a look again. I did test with one validator and then added few more which does seem to have change the behavior in the UI.

orangejenny commented:
@mkangia It may "work" but be a weird experience. I'm curious what the area I screenshotted would look like you if an import hit some standard row-level errors and also some of the new custom row-level errors.

mkangia committed:
removes redundant kwarg

mkangia committed:
updates to use specific error classes

- HQ groups errors by title, for better UI each error should be different class

mkangia committed:
avoids duplicate entries for a row

mkangia committed:
Merge pull request #30928 from dimagi/mk/samveg/count-limit-case-upload

[SAMVEG] limit case upload

mkangia committed:
replaces mandatory with required

mkangia commented on custom/samveg/case_importer/exceptions.py:
```diff
@@ -0,0 +1,2 @@
+class UnexpectedFileError(Exception):
```
This exception was added just to catch the unexpected file within samveg module. It is not to be used an an importer error.

I did try to make this work just like https://github.com/dimagi/commcare-hq/pull/30921/commits/f6e1ee67320ddb080b7250fcf78d4e483feb1ed7 but I could not succeed.
On file upload, in HQ, there is expected to be only a single error. But for extensions, there is the possibility of having multiple checks for file, so it seems better to return multiple errors messages and just join them later.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -59,9 +59,9 @@ def _validate_mandatory_columns(cls, row_num, raw_row):
                 missing_values.add(mandatory_column)
         if missing_values:
-            error_messages.append(_('Missing values for {column_names}').format(
+            error_messages.append(_('Mandatory columns {column_names}').format(
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/a7d29468b0a572d22b69e5090d95e61df752c16d

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -20,11 +20,11 @@ class MandatoryColumnsValidator(BaseValidator):
     def run(cls, spreadsheet):
         errors = []
-        errors.extend(cls._validate_missing_columns(spreadsheet))
+        errors.extend(cls._validate_mandatory_columns(spreadsheet))
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/a7d29468b0a572d22b69e5090d95e61df752c16d

mkangia committed:
adds one successful upload row test

- fixes comparison of date with date

mkangia committed:
fixes typo

mkangia committed:
updates documentation

mkangia committed:
calls run directly

mkangia committed:
removes redundant pass

mkangia committed:
updates documentation

mkangia committed:
uses removeprefix

mkangia committed:
uses ; to join messages

mkangia commented on custom/samveg/tests/case_importer/validators.py:
```diff
@@ -0,0 +1,124 @@
+        with get_spreadsheet(self.get_path('missing_columns_sncu_case_upload', 'xlsx')) as spreadsheet:
+            errors = MandatoryColumnsValidator.run(spreadsheet)
+        # extract missing columns from message
+        missing_columns = set(errors[0].split('Missing columns ')[1].split(', '))
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/b6a24ecf69ac0770c909b725b89834e8cd861979

mkangia commented on custom/samveg/tests/case_importer/validators.py:
```diff
@@ -0,0 +1,124 @@
+        with get_spreadsheet(self.get_path('missing_columns_rch_case_upload', 'xlsx')) as spreadsheet:
+            errors = MandatoryColumnsValidator.run(spreadsheet)
+        # extract missing columns from message
+        missing_columns = set(errors[0].split('Missing columns ')[1].split(', '))
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/b6a24ecf69ac0770c909b725b89834e8cd861979

mkangia commented on corehq/apps/case_importer/exceptions.py:
```diff
@@ -43,6 +43,11 @@ class InvalidCustomFieldNameException(ImporterError):
 
+class CustomImporterError(ImporterError):
+    """Raised for errors returned by extensions for file being imported"""
+    pass
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/0d4f1061b2760d877ff0e4f8618291ded2bb478b

mkangia commented on custom/samveg/commcare_extensions.py:
```diff
@@ -0,0 +1,39 @@
+    all_errors = []
+    for operation in row_level_operations:
+        if hasattr(operation, 'run'):
+            fields_to_update, errors = getattr(operation, 'run')(row_num, raw_row, fields_to_update)
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/178c9c1aa2baa98417ea98d33636378852a0a9f1

mkangia commented on custom/samveg/commcare_extensions.py:
```diff
@@ -0,0 +1,39 @@
+    errors = []
+    with case_upload.get_spreadsheet() as spreadsheet:
+        for validator in validators:
+            errors.extend(getattr(validator, 'run')(spreadsheet))
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/178c9c1aa2baa98417ea98d33636378852a0a9f1

mkangia commented on custom/samveg/tests/case_importer/validators.py:
```diff
@@ -68,3 +69,56 @@ def test_validate_mandatory_values(self):
         self._assert_missing_values_for_sheet('missing_values_sncu_case_upload', mandatory_values)
+
+
+class TestCallValidator(SimpleTestCase, TestFileMixin):
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/45b46232cbc28a5e52a87d3dd7bc27cc7791aafa

mkangia commented on corehq/apps/case_importer/extension_points.py:
```diff
@@ -0,0 +1,28 @@
+def custom_case_import_operations(domain, row_num, raw_row, fields_to_update):
+    """
+    Perform additional operations on a row and return final updates to be done.
+    To skip the row return error messages. These messages are then shown to the user.
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/c6639c79558e7ace765379d91fd9df2c645b724a

mkangia commented:
@orangejenny 
attached an image in the PR description to show the error messages look in the UI.

mkangia committed:
plug in additional operations

- restructure to have BaseOperation

mkangia committed:
adds new case properties

orangejenny commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -557,7 +557,8 @@ def add_error(self, row_num, error):
             self._errors[key][column_name]['rows'] = []
 
-        self._errors[key][column_name]['rows'].append(row_num)
+        if row_num not in self._errors[key][column_name]['rows']:
```
What happens if a row has multiple, different types of errors?

mkangia replied:
Then that row number would appear for each error, for example if you see row 2 in the screenshot in the description.

orangejenny commented on custom/samveg/tests/case_importer/validators.py:
```diff
@@ -58,7 +58,7 @@ def test_validate_required_columns(self):
 
 
-class TestRequiedValueValidator(SimpleTestCase, TestFileMixin):
+class TestRequiredValueValidator(SimpleTestCase, TestFileMixin):
```
👼

orangejenny reviewed (commented):
Updates look great. My only remaining question on the error UI is that it looks like the file-level error still displays the "Rows(s): ..." line, and like it might list out all rows in the file. That seems like it'll get unmanageable with large files.

mkangia commented on custom/samveg/case_importer/validators.py:
```diff
@@ -117,18 +127,14 @@ def run(cls, row_num, raw_row, fields_to_update, import_context):
-                        call_number=call_number
-                    )
-                )
+                error_messages.append(UploadLimitReachedError())
```
> It looks like the owner name is no longer displayed in the message, is that true?

Ya, I had to remove owner_name and call number.
Considering how the error is [added](https://github.com/dimagi/commcare-hq/blob/44f54201134f654a46ce05c6574b654600aad87e/corehq/apps/case_importer/do_import.py#L544), there is no real way to add information specific to a row in the error. 
The title of the error is used as the key. Then we have column_name and description. If we raise error with a message, that message is stored in the description but the description is updated each time a new error for the same key is added. So the description can be used if it stays the same for different errors, which I did for `RequiredValueValidator` in this [commit](https://github.com/dimagi/commcare-hq/pull/30921/commits/a52ab72473c0a851291225292ce948175b0a0892) by using mandatory columns instead of the missing values for the row.

mkangia replied:
I think we are looking at roughly 3k rows in an upload, for now.

when it comes to the upload limit, I think the user is not going to pay much attention, at least for this use case here. The concern is that up to 40 records should be uploaded, and not which records were rejected because of upload.

If we need a separate row for each owner, we would need to make an exception class with no title and then pass the message with the owner name in it, which would lead to a separate section for each owner_name in the UI, which just leads to a overall longer error message in the UI.
So I had to decide between that or a compact message in the UI, where I chose the latter because the former won't be useful but if we are asked to make a separate row for each owner on limit, we can do so.

mkangia commented:
> My only remaining question on the error UI is that it looks like the file-level error still displays the "Rows(s): ..." line, and like it might list out all rows in the file. That seems like it'll get unmanageable with large files.

File level errors happen at Step 1, and in that case the upload does not happen at all. If someone adds a column but leaves it all blank then what you are saying will happen. Hopefully no one would do that.

mkangia commented:
@orangejenny this is setup on staging on https://staging.commcarehq.org/a/samveg-quick/data/edit/import_cases/ in case you want to check the UI. There has been code changes, so the one reflecting the current state of code is the upload done on 12th. 
Feel free to download the corresponding file, modify it and upload it back.

mkangia committed:
updates to better naming

mkangia commented on custom/samveg/case_importer/exceptions.py:
```diff
@@ -0,0 +1,31 @@
+    pass
+
+
+class RequiredValueMissingError(CaseRowError):
```
https://github.com/dimagi/commcare-hq/pull/30921/commits/baa80818ca05332ac562471d5c42611f0728dd4c

mkangia committed:
Merge pull request #30938 from dimagi/mk/samveg/add-case-prop-case-upload

[SAMVEG] Add case prop case upload

millerdev reviewed (commented):
I didn't finish reviewing this yet, but I spend a lot of time trying to figure out the different types of validators and ran out of time today. Will try to wrap up my review tomorrow.

millerdev reviewed (changes requested):
Requesting changes. See review above.

millerdev commented on custom/samveg/case_importer/operations.py:
```diff
@@ -7,5 +12,6 @@ def run(cls, *args, **kwargs):
     @classmethod
     def run(cls, row_num, raw_row, fields_to_update):
-        # ToDo: Add new case property
+        today = datetime.utcnow().astimezone(pytz.timezone('Asia/Kolkata')).date()
```
Is this is doing what you want? This will return a different result depending on the local timezone of the machine where it is running. For reference, it's 8:27am on my computer.
```py
>>> utcnow = datetime.utcnow()  # NOTE UTC timestamp (not local time)
>>> zones = [
...   'no conversion',
...   None,
...   pytz.timezone('EST'),
...   pytz.timezone('UTC'),
...   pytz.timezone('Asia/Kolkata'),
... ]
>>> for zone in zones:
...     if zone == 'no conversion':
...         print(utcnow, zone)
...     else:
...         print(utcnow.astimezone(zone), 'as', zone)
... 
2022-01-14 13:27:02.969647 no conversion
2022-01-14 13:27:02.969647-05:00 as None
2022-01-14 13:27:02.969647-05:00 as EST
2022-01-14 18:27:02.969647+00:00 as UTC
2022-01-14 23:57:02.969647+05:30 as Asia/Kolkata
```
Note +10:30 difference between `utcnow` and the Asia/Kolkata time (it should only be +5:30).

I think you want something like this
```py
>>> utcnow_zoned = utcnow.replace(tzinfo=pytz.timezone('UTC'))
>>> utcnow_asia = utcnow_zoned.astimezone(pytz.timezone('Asia/Kolkata'))
>>> print(utcnow_zoned); print(utcnow_asia)
2022-01-14 13:27:02.969647+00:00
2022-01-14 18:57:02.969647+05:30
```
From the [`astimezone` docs](https://docs.python.org/3/library/datetime.html#datetime.datetime.astimezone):

> If called without arguments (or with `tz=None`) the system local timezone is assumed for the target timezone.

The fact that we assume any timezone-unaware timestamp is UTC leads to trip-ups like this. /rant

mkangia replied:
damn, thank you.

My intention here was to get the corresponding date that would be in `Asia/Kolkata` for when its run anywhere.
This would run on india.commcarehq.org which is hosted in India only, so I believe it would still get the right time.
But should definitely not rely on that, So this should be then datetime.now().astimezone(pytz.timezone('Asia/Kolkata')).date() then?

> The fact that we assume any timezone-unaware timestamp is UTC leads to trip-ups like this

Surely did

millerdev replied:
> So this should be then datetime.now().astimezone(pytz.timezone('Asia/Kolkata')).date() then?

That still depends on server's local timezone, which I don't like to depend on because it might change or be different on different servers. However, on re-reading the `datetime` docs, I realized that `datetime.now()` accepts a timezone argument. This is what you want:

`datetime.now(pytz.timezone('Asia/Kolkata')).date()`

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/30921/commits/fba72d1124f7fb8878b06c23fb911a6f2a76a5ac

millerdev reviewed (commented):
Review finished.

mkangia committed:
restructures operations

- unlink sheet validator from operations
- rename Operation to RowOperation
- have same signature for RowOperations

mkangia commented on custom/samveg/commcare_extensions.py:
```diff
@@ -0,0 +1,49 @@
+
+additional_row_level_operations = [
+    AddCustomCaseProperties,
+]
```
Thanks @millerdev . That does make the interface much cleaner and specific.

My intention originally was to just have "operations" irrespective of if they are validations or anything else and let the caller decide what they want to do with the return of each operation. In my effort to be flexible I might have made things more liberal.

Taking feedback from your changes here, I tried to align that structure better in https://github.com/dimagi/commcare-hq/pull/30921/commits/13cbca5b4054150ef47dfe01d472f840b2775e9d to bring more consistency to
1. what each operation (now role operation) would receive
2. what it should return
Added some documentation as well. Open to improvements.

Let me know your thoughts on this.
If it looks like that we can still benefit from individual base classes, `BaseRowValidator` and `BaseRowOperation`, I will be happy to add those, though there is only one `RowOperation`, so then taking from your feedback above in `BaseSheetValidator` that base class might not be needed anyway.

mkangia committed:
raises error if no response from extensions

mkangia committed:
updates to better error message

mkangia committed:
uses datetime.now to get correct date

millerdev commented on custom/samveg/case_importer/validators.py:
```diff
@@ -26,7 +26,7 @@
 
 
-class RequiredColumnsValidator(BaseOperation):
+class RequiredColumnsValidator(object):
```
nit: `(object)` is not needed since Python 3.
```suggestion
class RequiredColumnsValidator:
```

millerdev commented on custom/samveg/case_importer/validators.py:
```diff
@@ -103,9 +103,10 @@ def run(cls, row_num, raw_row, fields_to_update, **kwargs):
+class UploadLimitValidator(BaseRowOperation):
     @classmethod
-    def run(cls, row_num, raw_row, fields_to_update, import_context):
+    def run(cls, row_num, raw_row, fields_to_update, **kwargs):
+        import_context = kwargs['import_context']
```
I wouldn't call this an optional argument (as documented in `:param kwargs: Optional additional kwargs passed for a specific operation`). It will raise `KeyError` if it is not provided. If you want to make it required for this validator you could do something like
```py
def run(cls, row_num, raw_row, fields_to_update, *, import_context, **kwargs):
```
However, even that has a smell because the thing that calls `run(...)` always provides `import_context` and never provides any other arguments, so why not just make the signature for all operations require it?
```suggestion
    def run(cls, row_num, raw_row, fields_to_update, import_context):
```
It seems like you are planning for future (and as yet unknown) requirements by accepting `**kwargs`. This is an anti-pattern. Better to not make the code more flexible than it needs to be because that surprisingly often makes it harder to change in the future.

mkangia replied:
> It will raise KeyError if it is not provided

That is what I intended. 

I am almost done with changes I need for this feature and import_context is the only kwarg I need so makes sense to make that mandatory.
https://github.com/dimagi/commcare-hq/pull/30921/commits/f9da7eaa123af1ee707675fa2265b922c61be319

> It seems like you are planning for future (and as yet unknown) requirements by accepting **kwargs. This is an anti-pattern. Better to not make the code more flexible than it needs to be because that surprisingly often makes it harder to change in the future

Thanks for this note. A good reminder for me since I tend to do that.

mkangia committed:
Merge pull request #30958 from dimagi/mk/samveg/notify-unexpected-errors

[SAMVEG] notify unexpected errors

mkangia committed:
removes redundant inheritance

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

mkangia committed:
simplifies method signature

- make import_context an arg instead of optional kwarg

mkangia committed:
supports date type for calls

mkangia committed:
validates for at least one call column

mkangia commented on custom/samveg/commcare_extensions.py:
```diff
@@ -0,0 +1,18 @@
+    errors = []
+    with case_upload.get_spreadsheet() as spreadsheet:
+        for validator in validators:
+            errors.extend(getattr(validator, 'run')(spreadsheet))
```
thanks for the feedback Simon, This was done in review feedback before.

mkangia committed:
adds test for supported call types

mkangia committed:
Merge pull request #30968 from dimagi/mk/samveg/treat-calls-as-mandatory

[SAMVEG] Treat calls as mandatory

mkangia committed:
Merge pull request #30967 from dimagi/mk/samveg/support-parsed-date

[SAMVEG] Support parsed date
